src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Infer.java

```diff
@@ -651,6 +654,90 @@ public Type instantiateFunctionalInterface(DiagnosticPosition pos, Type funcInte
             return owntype;
         }
     }
+
+    public Type instantiatePatternType(Type expressionType, TypeSymbol patternTypeSymbol) {
+        if (expressionType.tsym == patternTypeSymbol)
+            return expressionType;
+
+        //step 1:
+        Type expressionTypeCaptured = types.capture(expressionType);
+        List<Type> params = patternTypeSymbol.type.allparams();
+        List<Type> capturedWildcards = List.nil();
+        //add synthetic captured ivars
+        for (Type ta : expressionTypeCaptured.getTypeArguments()) {
+            if (ta.hasTag(TYPEVAR) && ((TypeVar)ta).isCaptured()) {
+                params = params.prepend((TypeVar)ta);
+                capturedWildcards = capturedWildcards.prepend(ta);
+            }
+        }
+        InferenceContext c = new InferenceContext(this, params);
+        Type patternType = c.asUndetVar(patternTypeSymbol.type);
+        Type exprType = c.asUndetVar(expressionTypeCaptured);
+
+        capturedWildcards.forEach(s -> ((UndetVar) c.asUndetVar(s)).setNormal());
+
+        try {
+            //step 2:
+            if (exprType.isParameterized()) {
+                Type patternAsExpression =
+                        types.asSuper(patternType, exprType.tsym);
+                if (patternAsExpression == null ||
+                    !types.isSameType(patternAsExpression, exprType)) {
+                    return null;
+                }
+            }
+
+            doIncorporation(c, types.noWarnings);
+
+            //step 3:
+            List<Type> freshVars = instantiatePatternVars(params, c);
+
+            Type substituted = c.asInstType(patternTypeSymbol.type);
+
+            //step 4:
+            return types.upward(substituted, freshVars);
+        } catch (Infer.InferenceException ex) {
+            return null;
+        }
+    }
+
+    private List<Type> instantiatePatternVars(List<Type> vars, InferenceContext c) {
+        ListBuffer<Type> freshVars = new ListBuffer<>();
+        ListBuffer<Type> todo = new ListBuffer<>();
+
+        //step 1 - create fresh tvars
+        for (Type t : vars) {
```

Comment:
I'm not sure about this so just a question: shouldn't we run this loop until no more undetVars need to be instantiated? we could also limit until a maximum number of steps in case we find cases that doesn't converge to a solution

Reply:
yes I guess we should be fine and if we find any issue in the future we can always go back and react then